File: alexandria/src/main/java/com/alexandria/service/ReadingListService.java
package com.alexandria.service;

import com.alexandria.dto.document.DocumentSummary;
import com.alexandria.dto.readinglist.AddReadingListItemRequest;
import com.alexandria.dto.readinglist.CreateReadingListRequest;
import com.alexandria.dto.readinglist.ReadingListItemResponse;
import com.alexandria.dto.readinglist.ReadingListResponse;
import com.alexandria.dto.readinglist.ReadingListSummaryResponse;
import com.alexandria.dto.readinglist.UpdateReadingListRequest;
import com.alexandria.entity.Document;
import com.alexandria.entity.ReadingList;
import com.alexandria.entity.ReadingListItem;
import com.alexandria.entity.User;
import com.alexandria.entity.Visibility;
import com.alexandria.exception.DocumentNotFoundException;
import com.alexandria.exception.ReadingListItemAlreadyExistsException;
import com.alexandria.exception.ReadingListItemNotFoundException;
import com.alexandria.exception.ReadingListNotFoundException;
import com.alexandria.mapper.ReadingListMapper;
import com.alexandria.repository.DocumentRepository;
import com.alexandria.repository.ReadingListItemRepository;
import com.alexandria.repository.ReadingListRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Collectors;

@Transactional
@RequiredArgsConstructor
public class ReadingListService {

    private final ReadingListRepository readingListRepository;
    private final ReadingListItemRepository readingListItemRepository;
    private final DocumentRepository documentRepository;
    private final ReadingListMapper readingListMapper;
    private final InteractionService interactionService;

    @Transactional(readOnly = true)
    public Page<ReadingListSummaryResponse> getReadingLists(User currentUser, Pageable pageable) {
        return readingListRepository.findByUserId(currentUser.getId(), pageable)
                .map(readingListMapper::toSummaryResponse);
    }

    public ReadingListResponse createReadingList(CreateReadingListRequest request, User currentUser) {
        ReadingList list = new ReadingList();
        list.setName(request.name());
        list.setUser(currentUser);
        list.setCreatedAt(Instant.now());
        list.setItems(new ArrayList<>());
        ReadingList saved = readingListRepository.save(list);
        return readingListMapper.toResponse(saved, documentSummaries(saved));
    }

    @Transactional(readOnly = true)
    public ReadingListResponse getReadingList(UUID id) {
        ReadingList list = readingListRepository.findById(id)
                .orElseThrow(() -> new ReadingListNotFoundException(id));
        return readingListMapper.toResponse(list, documentSummaries(list));
    }

    public ReadingListResponse updateReadingList(UUID id, UpdateReadingListRequest request) {
        ReadingList list = readingListRepository.findById(id)
                .orElseThrow(() -> new ReadingListNotFoundException(id));
        list.setName(request.name());
        ReadingList saved = readingListRepository.save(list);
        return readingListMapper.toResponse(saved, documentSummaries(saved));
    }

    private Map<UUID, DocumentSummary> documentSummaries(ReadingList list) {
        List<ReadingListItem> items = list.getItems();
        if (items == null || items.isEmpty()) {
            return Map.of();
        }
        List<UUID> documentIds = items.stream()
                .map(item -> item.getDocument().getId())
                .toList();
        return documentRepository.findSummariesByIds(documentIds).stream()
                .collect(Collectors.toMap(DocumentSummary::id, Function.identity()));
    }

    public void deleteReadingList(UUID id) {
        ReadingList list = readingListRepository.findById(id)
                .orElseThrow(() -> new ReadingListNotFoundException(id));
        readingListRepository.delete(list);
    }

    public ReadingListItemResponse addItem(UUID listId, AddReadingListItemRequest request, UUID currentUserId) {
        ReadingList list = readingListRepository.findById(listId)
                .orElseThrow(() -> new ReadingListNotFoundException(listId));
        Document document = documentRepository.findById(request.documentId())
                .orElseThrow(() -> new DocumentNotFoundException(request.documentId()));
        if (document.getVisibility() == Visibility.PRIVATE
                && !document.getAuthor().getId().equals(currentUserId)) {
            throw new DocumentNotFoundException(request.documentId());
        }
        if (readingListItemRepository.findByReadingListIdAndDocumentId(listId, request.documentId()).isPresent()) {
            throw new ReadingListItemAlreadyExistsException(listId, request.documentId());
        }
        ReadingListItem item = new ReadingListItem();
        item.setReadingList(list);
        item.setDocument(document);
        item.setAddedAt(Instant.now());
        ReadingListItemResponse response = readingListMapper.toItemResponse(readingListItemRepository.save(item));
        interactionService.logBookmark(list.getUser(), document);
        return response;
    }

    public void removeItem(UUID listId, UUID documentId) {
        readingListRepository.findById(listId)
                .orElseThrow(() -> new ReadingListNotFoundException(listId));
        ReadingListItem item = readingListItemRepository.findByReadingListIdAndDocumentId(listId, documentId)
                .orElseThrow(() -> new ReadingListItemNotFoundException(listId, documentId));
        readingListItemRepository.delete(item);
    }
}

File: Alexandria-frontend/src/app/core/services/reading-list.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import type { Observable } from 'rxjs';
import type { SpringPage } from '../models/page.model';
import type {
  ReadingListResponse,
  ReadingListSummaryResponse,
  ReadingListItemResponse,
  CreateReadingListRequest,
  UpdateReadingListRequest,
  AddReadingListItemRequest,
} from '../models/reading-list.model';

@Injectable({ providedIn: 'root' })
export class ReadingListService {
  private readonly http = inject(HttpClient);
  private readonly base = `${environment.apiUrl}/reading-lists`;

  getReadingLists(page = 0, size = 20): Observable<SpringPage<ReadingListSummaryResponse>> {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<SpringPage<ReadingListSummaryResponse>>(this.base, { params });
  }

  createReadingList(request: CreateReadingListRequest): Observable<ReadingListResponse> {
    return this.http.post<ReadingListResponse>(this.base, request);
  }

  getReadingList(id: string): Observable<ReadingListResponse> {
    return this.http.get<ReadingListResponse>(`${this.base}/${id}`);
  }

  updateReadingList(id: string, request: UpdateReadingListRequest): Observable<ReadingListResponse> {
    return this.http.put<ReadingListResponse>(`${this.base}/${id}`, request);
  }

  deleteReadingList(id: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }

  addItem(id: string, request: AddReadingListItemRequest): Observable<ReadingListItemResponse> {
    return this.http.post<ReadingListItemResponse>(`${this.base}/${id}/items`, request);
  }

  removeItem(id: string, documentId: string): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}/items/${documentId}`);
  }
}
